src/object_store/src/object/opendal_engine/opendal_object_store.rs

```diff
@@ -195,34 +193,37 @@ impl ObjectStore for OpendalObjectStore {
 
 /// Store multiple parts in a map, and concatenate them on finish.
 pub struct OpenDalStreamingUploader {
-    op: Operator,
-    path: String,
-    buffer: BytesMut,
+    writer: Writer,
 }
 impl OpenDalStreamingUploader {
-    pub fn new(op: Operator, path: String) -> Self {
-        Self {
-            op,
-            path,
-            buffer: BytesMut::new(),
-        }
+    pub async fn new(op: Operator, path: String) -> ObjectResult<Self> {
+        let writer = op.writer(&path).await?;
+        Ok(Self { writer })
     }
 }
+
+unsafe impl Sync for OpenDalStreamingUploader {}
+
+const OPENDAL_BUFFER_SIZE: u64 = 8 * 1024 * 1024;
 #[async_trait::async_trait]
 impl StreamingUploader for OpenDalStreamingUploader {
     async fn write_bytes(&mut self, data: Bytes) -> ObjectResult<()> {
-        self.buffer.put(data);
+        self.writer.write(data).await?;
         Ok(())
     }
 
     async fn finish(mut self: Box<Self>) -> ObjectResult<()> {
-        self.op.write(&self.path, self.buffer).await?;
+        match self.writer.close().await {
+            Ok(_) => (),
+            Err(_) => self.writer.abort().await?,
```

Comment:
What happens to the partially uploaded data? Will it be deleted as well?

Reply:
No, OpenDAL has not implemented yet.

Reply:
:rofl: , nice catch! Tracked at https://github.com/apache/incubator-opendal/issues/2240

Reply:
abort has been implemented, will be released in next release.

Reply:
So fast, thanks!

Reply:
Will update this part after OpenDAL releasing,